### ISSUE TITLE
The `app_for_spawners` fixture - research about the implementation

### ISSUE THREAD

**Issue by GeorgianaElena** [jupyterhub/pytest-jupyterhub#25]
## Motivation
This issue **collects** details about the implementation of the `app_for_spawners` fixture that are otherwise scattered in other issues and PRs (most of the information here is already present in other places).

The purpose of this is to have everything in one place for context, and a tracker for the research work that has been done to come up to a solution.

It is part of a bigger task, that is tracked in https://github.com/jupyterhub/pytest-jupyterhub/issues/13.

## Iteration 1: Initial requirements
_Originally shared in https://github.com/jupyterhub/pytest-jupyterhub/issues/13_
- start by observing how the app fixtures are created and used by the [DockerSpawner](https://github.com/jupyterhub/dockerspawner/blob/8503af69161a3a543cc613f93ce7951ad30a1912/tests/conftest.py#L59-L100) and by [YarnSpawner](https://github.com/jupyterhub/yarnspawner/blob/1c40f529ddf9fd6217e9b8ae954fd669e4276c92/yarnspawner/tests/conftest.py#L21-L50)
- when writing the app_for_spawners fixture, try to design it so that, with proper configuration it could be used by both the DockerSpawner and the YarnSpawner instead of wha they are currently using

## Iteration 2: Refined focus areas based on the research in https://github.com/jupyterhub/pytest-jupyterhub/pull/19
_Originally shared in https://github.com/jupyterhub/pytest-jupyterhub/pull/19#pullrequestreview-1261453617_
- try not to include **spawner-specific** configuration (the configuration that makes sense for a specific spawner implementation, rather than for spawners **in general**)

- focus on finding ways to make anything that seems specific to a spawner implementation be configurable by a spawner in its own test suite.

- find common **JupyterHub-specific configuration** for the two spawners mentioned in the initial issue https://github.com/jupyterhub/pytest-jupyterhub/issues/13 (DockerSpawner and YarnSpawner)

- the DockerSpawner and the YarnSpawner code linked in the initial issue, provide two examples on hub customization. One, where the [hub is configured before the app is started](https://github.com/jupyterhub/yarnspawner/blob/1c40f529ddf9fd6217e9b8ae954fd669e4276c92/yarnspawner/tests/conftest.py#L23-L34) and another where the hub's config options are modified [while the hub application is running](https://github.com/jupyterhub/dockerspawner/blob/8503af69161a3a543cc613f93ce7951ad30a1912/tests/conftest.py#L70-L100).

- have `app_for_spawners` be the simplest hub app, that's responsible for starting and stopping the hub (like the one in the `jupyterhub` repo), but also for allowing a spawner to configure the **hub** 

- Feel free to add other helper fixtures and also to "copy" any other fixture from the two spawners to that seem to be helpful for a spawner **in general** and not just for that specific implementation (the copying part doesn't need to happen in this PR however)

## Iteration 3: Bottom-up approach

1. [ ] Implement the helper fixtures in https://github.com/jupyterhub/pytest-jupyterhub/issues/20
2. [ ] Implement the `app_for_spawners` fixture starting from these helper fixtures:
   1. `app_for_spawners` will be [requesting these fixtures](https://docs.pytest.org/en/6.2.x/fixture.html#fixtures-can-request-other-fixtures) and use them to configure the hub before starting the application.
       - _TIP_: the config coming from the two main helper fixtures might need to be [**merged**](https://traitlets.readthedocs.io/en/stable/config-api.html#traitlets.config.Config.merge) into a single Config object, that would then be passed to the hub application before starting it.
   3. the tests suite of the `dockerspawner` package for example, would use the `app_for_spawners` fixture of this pytest plugin, and only override these smaller fixtures to configure it as they wish.
       - To configure the hub spawner class, they would only provide a different implementation for `spawner_class` fixture and not touch `app_for_spawners`. Idelly there would be something like this:
      ```python
          from dockerspawner import DockerSpawner
          
          @pytest.fixture
          def spawner_class():
              """Allows tests to setup the Spawner class that will be used by the hub."""
              return DockerSpawner
          
          def test_something(app_for_spawners):
              assert app_for_spawners.something_else == something_else
      ```
      -  To configure various bits of the `DockerSpawner` object, they would only provide a different implementation for `spawner_config` fixture and not touch `app_for_spawners`. Idelly there would be something like this:
      ```python
          from dockerspawner import DockerSpawner
          from traitlets.config import Config
          
          @pytest.fixture
          def spawner_config():
            spawner_config = {
               "DockerSpawner": {
                  "prefix": ""dockerspawner-test",
                  "image": "test"
               }
            }
            return Config(spawner_config)
              
          
          def test_something(app_for_spawners):
              assert app_for_spawners.something_else == something_else
      ```

# Current ordered TODO list:
- [ ] 1. https://github.com/jupyterhub/pytest-jupyterhub/issues/20
- [ ] 2. Using the information in the section `Bottom-up approach` above, implement `app_for_spawners`
- [ ] 3. Document and test these fixtures as per https://github.com/jupyterhub/pytest-jupyterhub/issues/13

**Comment by GeorgianaElena** [comment]
@Sheila-nk, I opened this issue to keep all the information we have so far about how we envision the `"app_for_spawners"`, in one place. Kind of a summary of what was discussed so far.

We are currently at _Iteration 3: Bottom-up approach_. I've added more details there and expanded a bit more the goal of this approach in order to clarify a bit more what the outcome of https://github.com/jupyterhub/pytest-jupyterhub/issues/20 should be. 

Hopefully this last section also helps out answering your question in https://github.com/jupyterhub/pytest-jupyterhub/pull/24. But let's move the discussion about that work in that specific PR.

**Comment by Sheila-nk** [comment]
@GeorgianaElena thank you for opening this issue. It does help to have this as a reference for the entire vision and work towards the `app_for_spawners` fixture. 👏

**Comment by Sheila-nk** [comment]
For this `app_for_spawners` fixture, it seems we would have to explicitly mention to the users that the fixtures `spawner_class` and `spawner_config` need to be overridden?

**Comment by Sheila-nk** [comment]
Closing this issue as resolved by the merging of PR #36
